dbgap_extract.py

```diff
@@ -5,14 +5,111 @@
 import sys
 from datetime import datetime
 import logging
+import os
+import queue
 
 FILENAME = "extract-" + datetime.now().strftime("%m-%d-%Y-%H-%M-%S")
 REQUEST_URL = "https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/GetSampleStatus.cgi?study_id={}&rettype=xml"
+LOG_FILE = FILENAME + ".log"
+logging.basicConfig(filename=LOG_FILE, level=logging.DEBUG)
 
-logging.basicConfig(filename=FILENAME + ".log", level=logging.DEBUG)
-logging.getLogger().addHandler(logging.StreamHandler(sys.stdout))
-logging.getLogger("requests").setLevel(logging.WARNING)
-logging.getLogger("urllib3").setLevel(logging.WARNING)
+
+def setup_logging(log_filename):
+    global LOG_FILE
+    LOG_FILE = log_filename
+    if os.path.exists(LOG_FILE):
+        os.remove(LOG_FILE)
+    directory = os.path.dirname(log_filename)
+    if directory is not None and directory != "" and not os.path.exists(directory):
+        os.makedirs(directory)
+    fileh = logging.FileHandler(log_filename, "a")
+    log = logging.getLogger()
+    for hdlr in log.handlers[:]:
+        log.removeHandler(hdlr)
+    log.setLevel(logging.DEBUG)
+    log.addHandler(fileh)
+    log.addHandler(logging.StreamHandler(sys.stdout))
+    logging.getLogger("requests").setLevel(logging.WARNING)
+    logging.getLogger("urllib3").setLevel(logging.WARNING)
+
+
+def get_previous_version_of_study_accession(study_accession):
+    study_accession_split = study_accession.split(".")
+    version_number = int(study_accession_split[1][1:])
+    if version_number == 1:
+        # Can't go back any further
+        return -1
+    previous_version_of_study_accession = ".".join(
+        [
+            study_accession_split[0],
+            "v" + str(version_number - 1),
+            study_accession_split[2],
+        ]
+    )
+    return previous_version_of_study_accession
+
+
+def write_sample_rows_for_study(
+    study_accession, sample_elements, output_filename, field_names
+):
+    sample_rows_to_write_for_this_study = []
+    for sample in sample_elements:
+        try:
+            row = []
+            sample_dict = sample.attrib
+            uses = sample.findall("Uses")[0].findall("Use")
+            if len(uses) > 0:
+                uses_as_string = "; ".join(list(map(lambda x: x.text, uses)))
+                sample_dict["sample_use"] = uses_as_string
+            sra_datas = sample.findall("SRAData")[0].findall("Stats")
```

Comment:
I would prefer if this handling was split out into a different function since we may want to change how this gets represented later. and it would clean up this function a bit so it's not so nested